### PR TITLE
fix(idle): unrug withdraw underlying asset amounts

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR:

- ensures we *display* the actual underlying asset amounts in Idle withdraw, vs. considering the Idle wrapped asset as a 1/1 peg token currently.
- ensures we *broadcast* the matching Idle asset amounts in Idle withdraw

Also adds TODOs on how to consolidate this in a follow-up PR, so that we don't have to do back and forth calculations which make us lose precision, on top of the already lost precision by not using base uni amounts.

This does *not* tackle the issue we have across all DeFi modals currently WRT loss of precision, which will be solved in a follow-up PR by using base unit values across DeFi modals.

This also does not tackle the overview showing the underlying asset assuming 1/1 peg, to be tackled in a stacked PR.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/3392

## Testing

- Open an Idle opportunity you're staked in and click Withdraw
- Ensure the available to withdraw amount matches `amount of Idle token you have * (Idle token price / underlying token price)`
- Ensure Txs broadcast with the right Idle token amount to withdraw
- Ensure max withdraws don't error and the amount withdrawn is ~the full Idle token balance (off by remaining dust amounts, which is an issue we have across the DeFi section) - see https://etherscan.io/tx/0xa8ab3f0b9a404a65e5e2caa533227a2765fc9ec3cf749a8ac0c1411d0a855931 as a Max withdraw Tx

### Engineering

☝🏽 

### Operations

☝🏽 

#### Prod

image

#### This diff

image
image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3395** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3395" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/3395?utm_source=stack-comment).

File: src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx [hunk 1]
```diff
@@ -132,11 +142,18 @@ export const Confirm: React.FC<ConfirmProps> = ({ accountId, onNext }) => {
...
+        idleOpportunity.positionAsset.underlyingPerPosition,
+      )
+      // TODO: This is fine for now, but going forward we will need to:
+      // 1. Use base unit amounts everywhere, we shouldn't need to `.times(asset.precision)
+      // 2. Have a notion of a "display amount" and "underlying amount"
+      // These two notions should be decoupled and we should *not* have to calc back and forth from the wrapping and underlying assets
       const tx = await idleOpportunity.prepareWithdrawal({
         address: userAddress,
-        amount: bnOrZero(state.withdraw.cryptoAmount).times(`1e+${asset.precision}`).integerValue(),
+        amount: idleAssetWithdrawAmount.times(`1e+${asset.precision}`).integerValue(),
```
Comment (0xApotheosis) [inline src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx:156]: Isn't this just `toBaseUnit(idleAssetWithdrawAmount, asset.precision)`?

Comment (0xApotheosis) [inline src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx:156]: Isn't this just `toBaseUnit(idleAssetWithdrawAmount, asset.precision)`?

File: src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Withdraw.tsx [hunk 2]
```diff
@@ -73,29 +75,42 @@ export const Withdraw: React.FC<WithdrawProps> = ({ accountId, onNext }) => {
     selectEarnUserStakingOpportunity(state, opportunityDataFilter),
   )
 
+  useEffect(() => {
+    if (!opportunityData?.assetId) return
+    ;(async () => {
+      setIdleOpportunity(await idleInvestor.findByOpportunityId(opportunityData?.assetId))
```
Comment (0xApotheosis) [inline src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Withdraw.tsx:81]:
```suggestion
      setIdleOpportunity(await idleInvestor.findByOpportunityId(opportunityData.assetId))
```

File: src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx [hunk 3]
```diff
@@ -84,6 +86,14 @@ export const Confirm: React.FC<ConfirmProps> = ({ accountId, onNext }) => {
   const opportunityData = useAppSelector(state =>
     selectEarnUserStakingOpportunity(state, opportunityDataFilter),
   )
+
+  useEffect(() => {
+    if (!opportunityData?.assetId) return
+    ;(async () => {
+      setIdleOpportunity(await idleInvestor.findByOpportunityId(opportunityData?.assetId))
```
Comment (0xApotheosis) [inline src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx:93]:
```suggestion
      setIdleOpportunity(await idleInvestor.findByOpportunityId(opportunityData.assetId))
```

**Review by 0xApotheosis — DISMISSED** [review]
Tested locally, working as expected - was able to successfully withdraw the expected amount from idle.

A few trivial comments to tackle at your discretion.

File: src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx [hunk 4]
```diff
@@ -132,11 +142,18 @@ export const Confirm: React.FC<ConfirmProps> = ({ accountId, onNext }) => {
       )
         return
       dispatch({ type: IdleWithdrawActionType.SET_LOADING, payload: true })
-      const idleOpportunity = await idleInvestor.findByOpportunityId(opportunityData?.assetId)
       if (!idleOpportunity) throw new Error('No opportunity')
+
+      const idleAssetWithdrawAmount = bnOrZero(state.withdraw.cryptoAmount).div(
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx]: Definitely, nice catch!
As we start to move to base units as the source of truth everywhere, this is definitely not a nitpick but a naming that will avoid us future bugs and make the refactor to come easier 🤌 
49778dbccf1da0258255d63d977c50d17898bcaa

File: src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx (hunk 1 above)
Comment (gomesalexandre) [reply src/features/defi/providers/idle/components/IdleManager/Withdraw/components/Confirm.tsx]: Ab-solutely, and safer than `integerValue()` as well 🙏🏽 confirmed we're getting the same value, wrapped it into a BigNumber to honor `investor-idle` types

image
image

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**